### PR TITLE
Chapter 13: User microposts

### PR DESCRIPTION
**Pull request by baaohuii**

### CODE REVIEW

**Comment by baaohuii** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -155,3 +155,57 @@ input.btn.btn-primary {
...
+aside {
+  textarea {
+    height: 100px;
+    margin-bottom: 5px;
+  }
+}
+
+span.picture {
+  margin-top: 10px;
+  input {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  margin-top: 10px;

  input {
```

File: app/assets/stylesheets/microposts.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Microposts controller here.
```
Comment (chungpv-1008) [inline app/assets/stylesheets/microposts.scss]: remove this file

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -43,11 +43,11 @@ def get_user
...
-    def valid_user
-    return if @user && @user.activated? && @user.authenticated? (:reset, params[:id])
+  def valid_user
+    return if @user && @user.authenticated?(:reset, params[:id])
       
-    flash[:danger] = "Some messages"
-    redirect_to root_path
+      flash[:danger] = "Some messages"
+      redirect_to root_path
     end
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    flash[:danger] = "Some messages"
    redirect_to root_path
```

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -56,6 +56,5 @@ def check_expiration
     
       flash[:danger] = "Password reset has expired."
       redirect_to new_password_reset_path
-    end
-  end
+
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  end
```

File: app/controllers/static_page_controller.rb [hunk 5]
```diff
@@ -1,5 +1,10 @@
 class StaticPageController < ApplicationController
-  def home; end
+  def home
+    if logged_in?
+      @micropost = current_user.microposts.build
+      @feed_items = current_user.feed.paginate(page: params[:page])
```
Comment (chungpv-1008) [inline app/controllers/static_page_controller.rb]:
```suggestion
      @feed_items = current_user.feed.paginate page: params[:page]
```

File: app/helpers/microposts_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module MicropostsHelper
```
Comment (chungpv-1008) [inline app/helpers/microposts_helper.rb]: remove this file

File: app/models/micropost.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  has_one_attached :image
+  default_scope -> { order created_at: :desc }
```
Comment (chungpv-1008) [inline app/models/micropost.rb]: remove 'default_scope'. Note: Is never used by 'default_scope'

File: app/models/user.rb [hunk 8]
```diff
@@ -54,29 +60,40 @@ def activate
...
+
+    reset_sent_at < Settings.max_hours.hours.ago
+  end
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end
+  
+  def feed
+    Micropost.where("user_id = ?", id)
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
    Micropost.where user_id: id
```

File: app/views/microposts/_micropost.html.erb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+<li id="micropost-<%= micropost.id %>">
+  <%= link_to gravatar_for(micropost.user, size: 50), micropost.user %>
+  <span class="user"><%= link_to micropost.user.name, micropost.user %></span>
+  <span class="content"><%= micropost.content %>
+    <%= micropost.content %>
+    <%= image_tag micropost.display_image if micropost.image.attached? %>
+  </span>
+  <span class="timestamp">
+    Posted <%= time_ago_in_words(micropost.created_at) %> ago.
+    <% if current_user?(micropost.user) %>
```
Comment (chungpv-1008) [inline app/views/microposts/_micropost.html.erb]:
```suggestion
    Posted <%= time_ago_in_words micropost.created_at %> ago.
    <% if current_user? micropost.user %>
```

File: app/views/shared/_error_messages.html.erb [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(object.errors.count, "error") %>.
```
Comment (chungpv-1008) [inline app/views/shared/_error_messages.html.erb]:
```suggestion
      The form contains <%= pluralize(object.errors.size, "error") %>.
```

File: app/views/shared/_feed.html.erb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<% if @feed_items.any? %>
+  <ol class="microposts">
+    <%= render @feed_items %>
+  </ol>
+  <%= will_paginate @feed_items %>
```
Comment (chungpv-1008) [inline app/views/shared/_feed.html.erb]:
```suggestion
  <%= paginate @feed_items %>
```

File: app/views/shared/_micropost_form.html.erb [hunk 12]
```diff
@@ -0,0 +1,20 @@
+<%= form_for(@micropost) do |f| %>
```
Comment (chungpv-1008) [inline app/views/shared/_micropost_form.html.erb]:
```suggestion
<%= form_for @micropost do |f| %>
```

File: app/views/shared/_micropost_form.html.erb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+  <%= render 'shared/error_messages', object: f.object %>
```
Comment (chungpv-1008) [inline app/views/shared/_micropost_form.html.erb]: using " instead of '

File: app/views/shared/_micropost_form.html.erb [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+<% end %>
+
+<script type="text/javascript">
+  $("#micropost_image").bind("change", function() {
+    var size_in_megabytes = this.files[Settings.false].size/Settings.validations.photo.size/Settings.validations.photo.size;
+    if size_in_megabytes > Settings.validations.photo.max_size_Mb {
+      alert "Maximum file size is 5MB. Please choose a smaller file.";
+    }
+  });
+</script>
```
Comment (chungpv-1008) [inline app/views/shared/_micropost_form.html.erb]: Please move they to file .js

File: app/views/shared/_user_info.html.erb [hunk 15]
```diff
@@ -0,0 +1,4 @@
+<%= link_to gravatar_for(current_user, size: 50), current_user %>
+<h1><%= current_user.name %></h1>
+<span><%= link_to "view my profile", current_user %></span>
+<span><%= pluralize(current_user.microposts.count, "micropost") %></span>
```
Comment (chungpv-1008) [inline app/views/shared/_user_info.html.erb]:
```suggestion
<span><%= pluralize(current_user.microposts.size, "micropost") %></span>
```

File: app/views/users/show.html.erb [hunk 16]
```diff
@@ -8,4 +8,13 @@
       </h1>
     </section>
   </aside>
+  <div class="col-md-8">
+    <% if @user.microposts.any? %>
+      <h3><%= t "microposts.title" %> (<%= @user.microposts.count %>)</h3>
```
Comment (chungpv-1008) [inline app/views/users/show.html.erb]:
```suggestion
      <h3><%= t "microposts.title" %> (<%= @user.microposts.size %>)</h3>
```

**Comment by baaohuii** [comment]
ready

File: app/javascript/packs/image_upload.js [hunk 17]
```diff
@@ -0,0 +1,8 @@
+ $(function () {
+  $('#micropost_image').bind('change', function() {
+    var size_in_megabytes = this.files[0].size/1024/1024;
+    if (size_in_megabytes > 5) {
+      alert(I18n.t("shared.micropost_form.alert_detail"));
```
Comment (saobv-2223) [inline app/javascript/packs/image_upload.js:5]:
```suggestion
      alert(I18n.t("shared.micropost_form.alert_detail"));
      $(this).val(null);
```

**Comment by baaohuii** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 18]
```diff
@@ -43,19 +43,17 @@ def get_user
...
-    def valid_user
-    return if @user && @user.activated? && @user.authenticated? (:reset, params[:id])
+  def valid_user
+    return if @user && @user.authenticated?(:reset, params[:id])
       
-    flash[:danger] = "Some messages"
-    redirect_to root_path
-    end
+      flash[:danger] = "Some messages"
+      redirect_to root_path
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    flash[:danger] = "Some messages"
    redirect_to root_path
```

File: app/models/user.rb [hunk 19]
```diff
@@ -54,29 +60,40 @@ def activate
...
+
   end
 
   def send_password_reset_email
     UserMailer.password_reset(self).deliver_now
   end
 
   def password_reset_expired?
-    reset_sent_at < Setting.max_hours.hours.ago
+
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
```

**Comment by baaohuii** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**